loki/transform/transform_inline.py

```diff
@@ -229,88 +227,140 @@ def _map_unbound_dims(var, val):
+                )
+            arg_vars = tuple(v for v in callee_vars if v.name == arg.name)
+            argmap.update((v, _map_unbound_dims(v, qualified_value)) for v in arg_vars)
+        else:
+            argmap[arg] = val
+
+    # Deal with PRESENT check for optional arguments
+    present_checks = tuple(
+        check for check in FindInlineCalls().visit(callee.body) if check.function == 'PRESENT'
+    )
+    present_map = {
+        check: sym.Literal('.true.') if check.arguments[0] in call.arg_map else sym.Literal('.false.')
+        for check in present_checks
+    }
+    argmap.update(present_map)
+
+    # Recursive update of the map in case of nested variables to map
+    argmap = recursive_expression_map_update(argmap, max_iterations=10)
+
+    # Substitute argument calls into a copy of the body
+    callee_body = SubstituteExpressions(argmap, rebuild_scopes=True).visit(
+        callee.body.body, scope=caller
+    )
+
+    # Inline substituted body within a pair of marker comments
+    comment = Comment(f'! [Loki] inlined child subroutine: {callee.name}')
+    c_line = Comment('! =========================================')
+    return (comment, c_line) + as_tuple(callee_body) + (c_line, )
+
+
+def inline_subroutine_calls(routine, calls, callee, allowed_aliases=None):
+    """
+    Inline a set of call to an individual :any:`Subroutine` at source level.
+
+    This will replace all :any:`Call` objects to the specified
+    subroutine with an adjusted equivalent of the member routines'
+    body. For this, argument matching, including partial dimension
+    matching for array references is performed, and all
+    member-specific declarations are hoisted to the containing
+    :any:`Subroutine`.
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine in which to inline all calls to the member routine
+    calls : tuple or list of :any:`CallStatement`
+    callee : :any:`Subroutine`
+        The called target subroutine to be inlined in the parent
+    allowed_aliases : tuple or list of str or :any:`Expression`, optional
+        List of variables that will not be renamed in the parent scope, even
+        if they alias with a local declaration.
+    """
+    allowed_aliases = as_tuple(allowed_aliases)
+
+    # Ensure we process sets of calls to the same callee
+    assert all(call.routine == callee for call in calls)
+
+    # Prevent shadowing of callee's variables by renaming them a priori
     parent_variables = routine.variable_map
-    duplicate_locals = tuple(
-        v for v in member.variables
-        if v.name in parent_variables and v.name.lower() not in member._dummies
+    duplicates = tuple(
+        v for v in callee.variables
+        if v.name in parent_variables and v.name.lower() not in callee._dummies
     )
+    # Filter out allowed aliases to prevent suffixing
+    duplicates = tuple(v for v in duplicates if v.name not in allowed_aliases)
```

Comment:
Do we need to be careful here with regard to case-sensitivity, since comparing string `v.name`?

Reply:
Aha, yes. But a better fix is to use `v.symbol` as our base symbols do indeed inherit case-insensitive equivalence. :grin: I've added this to the base test.